wip/value-domain.md

```diff
@@ -0,0 +1,128 @@
+# Rust Value Domain
+
+**Note:** This document is not normative nor endorsed by the UCG WG. Its purpose is to be the basis for discussion and to set down some key terminology.
+
+The purpose of this document is to describe what the set of *all possible values* is in Rust.
+This is an important definition: one key element of a Rust specification will be to define the [representation relation][representation] of every type.
+This relation relates values with lists of bytes: it says, for a given value and list of bytes, if that value is represented by that list.
+However, before we can even start specifying the relation, we have to specify the involved domains.
+`Byte` is defined as part of [the memory interface][memory-interface]; this document is about defining `Value`.
+
+[representation]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/glossary.md#representation
+[memory-interface]: memory-interface.md
+
+## Values
+
+The Rust value domain is described by the following (incomplete) type definition (in terms of an arbitrary `Pointer` type instantiated by [the memory interface][memory-interface]):
+
+```rust
+enum Value<Pointer> {
+    /// A mathematical integer.
+    Int(BigInt),
+    /// A Boolean value.
+    Bool(bool),
```

Comment:
Ostensibly it would be more precise to define `Int` and `Bool` using a sigma type with `({x: Nat | x ∈ start..end}, start: Nat, end: Nat)` or using [`Fin : (n : Nat) -> Type`](https://github.com/idris-lang/Idris-dev/blob/master/libs/base/Data/Fin.idr#L11) but well... can't use Rust for notation anymore. =P

Reply:
I also don't think I agree. If this was Coq I'd use `Z`, not some `Fin` type, for `Int`.

Reply:
Ah I see. That makes sense, the range would be more appropriate for the value relation for a type instead. Tho... in that case, why do we need `Bool` at all and why is that not subsumed by `Int`?

Reply:
Are you a set theorist or what? ;) "boolean" and "integer" are two distinct mathematical concepts.

Reply:
Sure, but why is this distinction made here rather than when relating values to `bool`?

Reply:
Because the entire point of `Value` is to capture the mathematical concepts that are represented by the data we store in memory. If there is no concept of a "boolean" in `Value`, then the language literally does not have the concept of a "boolean" anywhere.